src/codegen/ast_interpreter.cpp

```diff
@@ -1010,33 +1054,21 @@ Value ASTInterpreter::visit_str(AST_Str* node) {
 }
 
 Value ASTInterpreter::visit_name(AST_Name* node) {
-    switch (node->lookup_type) {
-        case AST_Name::UNKNOWN: {
-            ScopeInfo::VarScopeType vst = scope_info->getScopeTypeOfName(node->id);
-            if (vst == ScopeInfo::VarScopeType::GLOBAL) {
-                node->lookup_type = AST_Name::GLOBAL;
-                return getGlobal(source_info->parent_module, &node->id.str());
-            } else if (vst == ScopeInfo::VarScopeType::DEREF) {
-                node->lookup_type = AST_Name::CLOSURE;
-                return getattr(passed_closure, node->id.c_str());
-            } else {
-                bool is_old_local = (vst == ScopeInfo::VarScopeType::NAME);
-                node->lookup_type = is_old_local ? AST_Name::LOCAL : AST_Name::FAST_LOCAL;
-
-                SymMap::iterator it = sym_table.find(node->id);
-                if (it != sym_table.end()) {
-                    Box* value = it->second;
-                    return value;
-                }
-
-                // classdefs (and some other cases like eval) have different scoping rules than functions:
-                if (is_old_local)
-                    return getGlobal(source_info->parent_module, &node->id.str());
-
-                assertNameDefined(0, node->id.c_str(), UnboundLocalError, true);
-                return Value();
-            }
+    if (node->lookup_type == AST_Name::UNKNOWN) {
+        ScopeInfo::VarScopeType vst = scope_info->getScopeTypeOfName(node->id);
+        if (vst == ScopeInfo::VarScopeType::GLOBAL) {
+            node->lookup_type = AST_Name::GLOBAL;
```

Comment:
Sorry to harp on this, but I really think we can get rid of the duplicate system and that that would reduce the mental burden.  There's still a mapping from `VarScopeType` to the way that the AST interpreter retrieves the value for the name; the only thing is that we have the optimization that closure variables also got set in the local symbol table so we can receive them from there.  So we could just have the `FAST and `CLOSURE` cases fall through to the same code that accesses them in the symbol table.